File: tutorials/b3-running-cxx/run-cmaes.cpp
#include "model/heat2d.h"
#include "korali.hpp"

int main(int argc, char* argv[])
{
 auto k = korali::Engine();
 auto p = heat2DInit(&argc, &argv);

 k["Problem"]["Type"] = "Evaluation/Bayesian/Inference/Reference";
 k["Problem"]["Likelihood Model"] = "Additive Normal";
 k["Problem"]["Reference Data"] = p.refTemp;
 k["Problem"]["Computational Model"] = &heat2DSolver;

 k["Solver"]["Type"] = "Optimizer/CMAES";
 k["Solver"]["Population Size"] = 32;
 k["Solver"]["Termination Criteria"]["Max Generations"] = 20;

 k["Variables"][0]["Name"] = "Intensity";
 k["Variables"][0]["Bayesian Type"] = "Computational";
 k["Variables"][0]["Prior Distribution"]["Type"] = "Univariate/Uniform";
 k["Variables"][0]["Prior Distribution"]["Minimum"] = 10.0;
 k["Variables"][0]["Prior Distribution"]["Maximum"] = 60.0;
 k["Variables"][0]["Initial Mean"] = 30.0;
 k["Variables"][0]["Initial Standard Deviation"] = 5.0;

 k["Variables"][1]["Name"] = "PosX";
 k["Variables"][1]["Bayesian Type"] = "Computational";
 k["Variables"][1]["Prior Distribution"]["Type"] = "Univariate/Uniform";
 k["Variables"][1]["Prior Distribution"]["Minimum"] = 0.0;
 k["Variables"][1]["Prior Distribution"]["Maximum"] = 0.5;
 k["Variables"][1]["Initial Mean"] = 0.25;
 k["Variables"][1]["Initial Standard Deviation"] = 0.01;

 k["Variables"][2]["Name"] = "PosY";
 k["Variables"][2]["Bayesian Type"] = "Computational";
 k["Variables"][2]["Prior Distribution"]["Type"] = "Univariate/Uniform";
 k["Variables"][2]["Prior Distribution"]["Minimum"] = 0.6;
 k["Variables"][2]["Prior Distribution"]["Maximum"] = 1.0;
 k["Variables"][2]["Initial Mean"] = 0.8;
 k["Variables"][2]["Initial Standard Deviation"] = 0.1;

 k["Variables"][3]["Name"] = "Sigma";
 k["Variables"][3]["Bayesian Type"] = "Statistical";
 k["Variables"][3]["Prior Distribution"]["Type"] = "Univariate/Uniform";
 k["Variables"][3]["Prior Distribution"]["Minimum"] = 0.0;
 k["Variables"][3]["Prior Distribution"]["Maximum"] = 20.0;
 k["Variables"][3]["Initial Mean"] = 10.0;
 k["Variables"][3]["Initial Standard Deviation"] = 1.0;

 k.run();

 return 0;
}

File: examples/study.cases/RBCStretch/run-cmaes.py
#!/usr/bin/env python3
import sys
sys.path.append("model")
import data
import model
import korali

k = korali.Engine()
e = korali.Experiment()

# Defining Concurrent Jobs
cJobs = 1
if (len(sys.argv) > 1):
  cJobs = int(sys.argv[1])

# Setting up the reference likelihood for the Bayesian Problem
e["Problem"]["Type"] = "Bayesian/Reference"
e["Problem"]["Likelihood Model"] = "Normal"
e["Problem"]["Reference Data"] = data.getReferenceData().tolist()
e["Problem"]["Computational Model"] = lambda koraliData: model.evaluate(
    koraliData,
    data.getReferencePoints().tolist(), k.getMPIComm())

e["Distributions"][0]["Name"] = "Uniform 0"
e["Distributions"][0]["Type"] = "Univariate/Uniform"
e["Distributions"][0]["Minimum"] = 0.2
e["Distributions"][0]["Maximum"] = 0.6

e["Distributions"][1]["Name"] = "Uniform 1"
e["Distributions"][1]["Type"] = "Univariate/Uniform"
e["Distributions"][1]["Minimum"] = 10.0
e["Distributions"][1]["Maximum"] = 40.0

e["Distributions"][2]["Name"] = "Uniform 2"
e["Distributions"][2]["Type"] = "Univariate/Uniform"
e["Distributions"][2]["Minimum"] = 1e-5
e["Distributions"][2]["Maximum"] = 2.0

# Setting up Problem Variables
e["Variables"][0]["Name"] = "x0"
e["Variables"][0]["Prior Distribution"] = "Uniform 0"
e["Variables"][0]["Initial Mean"] = 0.4
e["Variables"][0]["Initial Standard Deviation"] = 0.05

e["Variables"][1]["Name"] = "ks"
e["Variables"][1]["Prior Distribution"] = "Uniform 1"
e["Variables"][1]["Initial Mean"] = 25.0
e["Variables"][1]["Initial Standard Deviation"] = 3.0

e["Variables"][2]["Name"] = "[Sigma]"
e["Variables"][2]["Prior Distribution"] = "Uniform 2"
e["Variables"][2]["Initial Mean"] = 1.0
e["Variables"][2]["Initial Standard Deviation"] = 0.05

# Setting CMAES as Solver
e["Solver"]["Type"] = "CMAES"
e["Solver"]["Population Size"] = int(cJobs)
e["Solver"]["Termination Criteria"]["Max Generations"] = 50

# General Settings
e["Console Output"]["Verbosity"] = "Detailed"
e["File Output"]["Path"] = "_result_rbc_cmaes_" + str(cJobs).zfill(4) + "_nodes"

# Configuring Distributed Conduit
k["Conduit"]["Type"] = "Distributed"
k["Conduit"]["Workers Per Team"] = 1
k.run(e)
